Include/datetime.h

```diff
@@ -119,39 +119,168 @@ typedef struct
 // o is a pointer to a time or a datetime object.
 #define _PyDateTime_HAS_TZINFO(o)  (((_PyDateTime_BaseTZInfo *)(o))->hastzinfo)
 
-#define PyDateTime_GET_YEAR(o)     ((((PyDateTime_Date*)(o))->data[0] << 8) | \
-                     ((PyDateTime_Date*)(o))->data[1])
-#define PyDateTime_GET_MONTH(o)    (((PyDateTime_Date*)(o))->data[2])
-#define PyDateTime_GET_DAY(o)      (((PyDateTime_Date*)(o))->data[3])
-
-#define PyDateTime_DATE_GET_HOUR(o)        (((PyDateTime_DateTime*)(o))->data[4])
-#define PyDateTime_DATE_GET_MINUTE(o)      (((PyDateTime_DateTime*)(o))->data[5])
-#define PyDateTime_DATE_GET_SECOND(o)      (((PyDateTime_DateTime*)(o))->data[6])
-#define PyDateTime_DATE_GET_MICROSECOND(o)              \
-    ((((PyDateTime_DateTime*)(o))->data[7] << 16) |       \
-     (((PyDateTime_DateTime*)(o))->data[8] << 8)  |       \
-      ((PyDateTime_DateTime*)(o))->data[9])
-#define PyDateTime_DATE_GET_FOLD(o)        (((PyDateTime_DateTime*)(o))->fold)
-#define PyDateTime_DATE_GET_TZINFO(o)      (_PyDateTime_HAS_TZINFO((o)) ? \
-    ((PyDateTime_DateTime *)(o))->tzinfo : Py_None)
+#define _PyDateTime_Date_CAST(op) ((PyDateTime_Date*)(op))
+
+static inline int PyDateTime_GET_YEAR(PyObject *o) {
+    PyDateTime_Date *date = _PyDateTime_Date_CAST(o);
+    return ((date->data[0] << 8) | date->data[1]);
+}
+#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 < 0x030c0000
```

Comment:
It might be worth a comment somewhere explaining why these macros exist, or just a reference to the relevant section in PEP 670. I suspect that future maintainers will be confused about why this exists.

Though it's not clear what the right balance is between repetition and clarity. One comment on the first one of these macros? A comment on every `#if` line? I'm inclined to just do one comment at the beginning in each file where this sort of thing is defined.